File: sdk/src/hd.ts
/**
 * Hierarchical-deterministic (HD) wallets for SOV.
 *
 * SOV signs with Ed25519, so HD follows the Ed25519 standards rather than
 * BIP-32's secp256k1 elliptic-curve math:
 *
 *   - **BIP-39** turns a human mnemonic phrase into a 512-bit seed
 *     (PBKDF2-HMAC-SHA512, 2048 rounds), via the audited `@scure/bip39`.
 *   - **SLIP-0010** derives a tree of Ed25519 keys from that seed using
 *     HMAC-SHA512 (the audited `@noble/hashes` primitive). Ed25519 supports
 *     *hardened* derivation only, so every path component is hardened.
 *   - **BIP-44** gives the path layout: `m / 44' / coin' / account' / change' /
 *     index'`.
 *
 * The 32-byte key SLIP-0010 produces at a leaf *is* an Ed25519 seed, so it feeds
 * directly into {@link Keypair.fromSeed} — and the resulting public key is the
 * one bound to a named account on-chain (an account's controlling / access key:
 * see `chain/crates/state/src/account.rs` `Account.key` and the runtime's
 * authorization check in `chain/crates/runtime/src/execution.rs`). One mnemonic
 * therefore deterministically controls a whole family of named SOV accounts.
 *
 * SCHEME NOTE: {@link HdWallet.deriveKeypair} derives the **classical Ed25519**
 * key (scheme `0x00`), a valid on-chain account key. SOV's *default* and
 * recommended scheme is the **hybrid Ed25519 + ML-DSA-65** post-quantum key —
 * use {@link HdWallet.deriveHybridKeypair}, which expands the identical
 * SLIP-0010 leaf via `@noble/post-quantum` and is byte-for-byte identical to the
 * node's `sov-wallet import` / `Keypair::hybrid_from_seed` (pinned by the
 * cross-impl KAT in `test/hybrid.test.ts`).
 *
 * No derivation logic is hand-rolled beyond SLIP-0010's HMAC chaining; the hash
 * primitives are audited libraries, and correctness is pinned against the
 * SLIP-0010 specification's official Ed25519 test vectors in `test/hd.test.ts`.
 */

import { hmac } from "@noble/hashes/hmac.js";
import { sha512 } from "@noble/hashes/sha512.js";
import {
  generateMnemonic as scureGenerateMnemonic,
  mnemonicToSeedSync,
  validateMnemonic as scureValidateMnemonic,
} from "@scure/bip39";
import { wordlist } from "@scure/bip39/wordlists/english";

import { Keypair, KeyError } from "./keys.js";
import { HybridKeypair } from "./hybrid.js";

/** The hardened-derivation offset, 2^31. SLIP-0010 Ed25519 requires every
 * path component to be hardened, i.e. `>= HARDENED_OFFSET`. */
export const HARDENED_OFFSET = 0x8000_0000;

/** BIP-44 purpose constant. */
const PURPOSE = 44;

/**
 * SOV's BIP-44 coin type.
 *
 * PROVISIONAL: SOV is not (yet) registered in SLIP-0044, so this is a chosen,
 * fixed value rather than an assigned slot. It encodes the ASCII bytes `"SOV"`
 * (0x534F56) so it is memorable and self-documenting, and is `< 2^31` so it is a
 * valid hardened path index. It is fixed here so every SOV wallet derives the
 * same addresses; changing it would fork all derivation paths, so treat it as
 * stable and only revisit under a coordinated migration if a SLIP-0044 slot is
 * formally assigned.
 */
export const SOV_COIN_TYPE = 0x53_4f_56; // "SOV" — provisional, not yet SLIP-0044 registered

/** SLIP-0010 master HMAC key for the Ed25519 curve. */
const ED25519_CURVE = new TextEncoder().encode("ed25519 seed");

/** A node in the SLIP-0010 key tree: a 32-byte key and its 32-byte chain code. */
interface Slip10Node {
  /** The 32-byte derived key. At a leaf this is an Ed25519 seed. */
  readonly key: Uint8Array;
  /** The 32-byte chain code carried to children. */
  readonly chainCode: Uint8Array;
}

/** Big-endian 4-byte serialization of a u32 index (SLIP-0010 `ser32`). */
function ser32(index: number): Uint8Array {
  const out = new Uint8Array(4);
  out[0] = (index >>> 24) & 0xff;
  out[1] = (index >>> 16) & 0xff;
  out[2] = (index >>> 8) & 0xff;
  out[3] = index & 0xff;
  return out;
}

/** Derive the SLIP-0010 Ed25519 master node from a seed. */
function slip10Master(seed: Uint8Array): Slip10Node {
  const i = hmac(sha512, ED25519_CURVE, seed);
  return { key: i.slice(0, 32), chainCode: i.slice(32, 64) };
}

/**
 * SLIP-0010 Ed25519 hardened child derivation. Unlike BIP-32/secp256k1, the
 * child key is `I_L` directly (no scalar addition of the parent key), and only
 * hardened indices are defined.
 */
function slip10CkdPriv(parent: Slip10Node, index: number): Slip10Node {
  if (index < HARDENED_OFFSET) {
    throw new KeyError(
      `SLIP-0010 Ed25519 supports hardened derivation only; index ${index} is not hardened`,
    );
  }
  // data = 0x00 || ser256(k_par) || ser32(index)
  const data = new Uint8Array(1 + 32 + 4);
  data[0] = 0x00;
  data.set(parent.key, 1);
  data.set(ser32(index), 33);
  const i = hmac(sha512, parent.chainCode, data);
  return { key: i.slice(0, 32), chainCode: i.slice(32, 64) };
}

/** Walk a path of (hardened) indices from the master node derived from `seed`. */
function derivePath(indices: readonly number[], seed: Uint8Array): Slip10Node {
  let node = slip10Master(seed);
  for (const index of indices) {
    node = slip10CkdPriv(node, index);
  }
  return node;
}

/**
 * Parse a SLIP-0010 path string like `m/44'/5459798'/0'/0'/0'` into hardened
 * indices. Every component must be hardened (suffix `'`, `h`, or `H`), because
 * Ed25519 derivation defines only hardened children. Throws {@link KeyError}.
 */
export function parsePath(path: string): number[] {
  const parts = path.trim().split("/");
  if (parts[0] !== "m") {
    throw new KeyError(`HD path must start with "m": ${path}`);
  }
  return parts.slice(1).map((p) => {
    const hardened = p.endsWith("'") || p.endsWith("h") || p.endsWith("H");
    if (!hardened) {
      throw new KeyError(
        `SLIP-0010 Ed25519 requires hardened path components; "${p}" is not (append ')`,
      );
    }
    const numStr = p.slice(0, -1);
    const n = Number(numStr);
    if (!Number.isInteger(n) || n < 0 || n >= HARDENED_OFFSET || !/^\d+$/.test(numStr)) {
      throw new KeyError(`invalid HD path component: "${p}"`);
    }
    return n + HARDENED_OFFSET;
  });
}

/**
 * The standard SOV BIP-44 derivation indices for a given account and address
 * index: `m / 44' / SOV_COIN_TYPE' / account' / 0' / index'` (all hardened).
 */
export function sovPath(account = 0, index = 0): number[] {
  assertIndex("account", account);
  assertIndex("index", index);
  return [
    PURPOSE + HARDENED_OFFSET,
    SOV_COIN_TYPE + HARDENED_OFFSET,
    account + HARDENED_OFFSET,
    0 + HARDENED_OFFSET,
    index + HARDENED_OFFSET,
  ];
}

/** Human-readable form of {@link sovPath}, e.g. `m/44'/5459798'/0'/0'/0'`. */
export function sovPathString(account = 0, index = 0): string {
  assertIndex("account", account);
  assertIndex("index", index);
  return `m/${PURPOSE}'/${SOV_COIN_TYPE}'/${account}'/0'/${index}'`;
}

function assertIndex(label: string, n: number): void {
  if (!Number.isInteger(n) || n < 0 || n >= HARDENED_OFFSET) {
    throw new KeyError(`${label} must be an integer in [0, 2^31); got ${n}`);
  }
}

/**
 * Generate a fresh BIP-39 mnemonic from secure entropy (default 24 words /
 * 256-bit). The audited `@scure/bip39` draws from `crypto.getRandomValues`.
 */
export function generateMnemonic(strength: 128 | 160 | 192 | 224 | 256 = 256): string {
  return scureGenerateMnemonic(wordlist, strength);
}

/** Whether `mnemonic` is a valid BIP-39 phrase (wordlist + checksum). */
export function validateMnemonic(mnemonic: string): boolean {
  return scureValidateMnemonic(mnemonic, wordlist);
}

/**
 * A hierarchical-deterministic SOV wallet: one BIP-39 seed from which an entire
 * tree of Ed25519 {@link Keypair}s is derived deterministically. The seed is
 * held privately and never serialized.
 */
export class HdWallet {
  private readonly seed: Uint8Array;

  private constructor(seed: Uint8Array) {
    this.seed = Uint8Array.from(seed);
  }

  /**
   * Build a wallet from a BIP-39 mnemonic (optionally salted by a passphrase —
   * the BIP-39 "25th word"). Rejects a phrase that fails the wordlist/checksum.
   */
  static fromMnemonic(mnemonic: string, passphrase = ""): HdWallet {
    if (!validateMnemonic(mnemonic)) {
      throw new KeyError("invalid BIP-39 mnemonic (wordlist or checksum)");
    }
    return new HdWallet(mnemonicToSeedSync(mnemonic.normalize("NFKD"), passphrase));
  }

  /**
   * Build a wallet directly from a raw BIP-39 seed (>= 16 bytes), for callers
   * that already hold seed material. Most users should use {@link fromMnemonic}.
   */
  static fromSeed(seed: Uint8Array): HdWallet {
    if (seed.length < 16) {
      throw new KeyError(`HD seed must be at least 16 bytes, got ${seed.length}`);
    }
    return new HdWallet(seed);
  }

  /** Derive the keypair at an explicit SLIP-0010 path (all components hardened). */
  deriveKeypairAtPath(path: string | readonly number[]): Keypair {
    const indices = typeof path === "string" ? parsePath(path) : path;
    const node = derivePath(indices, this.seed);
    return Keypair.fromSeed(node.key);
  }

  /**
   * Derive the **classical Ed25519** keypair for a SOV `account` and address
   * `index` on the standard BIP-44 path `m/44'/SOV_COIN_TYPE'/account'/0'/index'`.
   * A valid on-chain key; for SOV's default post-quantum scheme use
   * {@link deriveHybridKeypair}.
   */
  deriveKeypair(account = 0, index = 0): Keypair {
    return this.deriveKeypairAtPath(sovPath(account, index));
  }

  /**
   * The raw 32-byte SLIP-0010 **leaf seed** for an `account`/`index` — the
   * deterministic, scheme-agnostic HD output (identical to the node's
   * `HdWallet::derive_seed`). Feed it to either key scheme.
   */
  deriveSeed(account = 0, index = 0): Uint8Array {
    return Uint8Array.from(derivePath(sovPath(account, index), this.seed).key);
  }

  /** Derive the hybrid post-quantum keypair at an explicit SLIP-0010 path. */
  deriveHybridKeypairAtPath(path: string | readonly number[]): HybridKeypair {
    const indices = typeof path === "string" ? parsePath(path) : path;
    return HybridKeypair.fromSeed(derivePath(indices, this.seed).key);
  }

  /**
   * Derive the **hybrid Ed25519 + ML-DSA-65** (post-quantum) keypair for a SOV
   * `account`/`index` on the standard BIP-44 path — SOV's default scheme, and
   * byte-for-byte identical to the node's `sov-wallet import`. The same mnemonic
   * restores the same hybrid key, every time.
   */
  deriveHybridKeypair(account = 0, index = 0): HybridKeypair {
    return this.deriveHybridKeypairAtPath(sovPath(account, index));
  }

  /** Expose the raw BIP-39 seed bytes (handle with care; secret material). */
  exportSeed(): Uint8Array {
    return Uint8Array.from(this.seed);
  }
}

File: chain/crates/wallet/src/hd.rs
//! Hierarchical-deterministic (HD) wallets — Rust side, at parity with the SDK
//! (`sdk/src/hd.ts`).
//!
//! SOV signs with Ed25519, so HD follows the Ed25519 standards rather than
//! BIP-32's secp256k1 math:
//!
//!   - **BIP-39** turns a mnemonic into a 512-bit seed (PBKDF2-HMAC-SHA512),
//!     via the audited `bip39` crate;
//!   - **SLIP-0010** derives a tree of Ed25519 keys from that seed using
//!     HMAC-SHA512 (`hmac` + `sha2`). Ed25519 supports *hardened* derivation
//!     only, so every path component is hardened;
//!   - **BIP-44** gives the path `m / 44' / SOV_COIN_TYPE' / account' / 0' /
//!     index'`.
//!
//! The 32-byte key SLIP-0010 produces at a leaf is the deterministic **leaf
//! seed**. SOV expands it into the account's signing key: the default is the
//! **hybrid Ed25519 + ML-DSA-65** post-quantum key ([`Keypair::hybrid_from_seed`]
//! / [`HdWallet::derive_keypair`]), with a legacy classical Ed25519 key
//! ([`HdWallet::derive_ed25519_keypair`]) also available. The leaf-seed
//! derivation is pinned byte-for-byte to the SDK by a cross-implementation
//! known-answer test, so the same mnemonic restores the same SOV wallet in Rust
//! and TypeScript (the SDK currently expands the leaf to the Ed25519 key; its
//! hybrid expansion lands with ML-DSA-65 in TypeScript).

use bip39::Mnemonic;
use hmac::{Hmac, Mac};
use sha2::Sha512;
use sov_crypto::Keypair;

type HmacSha512 = Hmac<Sha512>;

/// The hardened-derivation offset (2^31). SLIP-0010 Ed25519 derivation is
/// hardened-only, so every path component carries this bit.
const HARDENED: u32 = 0x8000_0000;

/// BIP-44 purpose.
const PURPOSE: u32 = 44;

/// SOV's BIP-44 coin type — the ASCII bytes `"SOV"` (`0x53_4f_56` = 5_459_798).
/// PROVISIONAL: SOV is not yet SLIP-0044 registered; fixed here so every wallet
/// derives the same addresses. Must equal `SOV_COIN_TYPE` in `sdk/src/hd.ts`.
pub const SOV_COIN_TYPE: u32 = 0x53_4f_56;

/// SLIP-0010 master HMAC key for the Ed25519 curve.
const ED25519_CURVE: &[u8] = b"ed25519 seed";

/// Errors building an HD wallet.
#[derive(Debug, thiserror::Error)]
pub enum HdError {
    /// The mnemonic failed BIP-39 validation (wordlist or checksum).
    #[error("invalid BIP-39 mnemonic: {0}")]
    Mnemonic(String),
}

/// Generate a fresh BIP-39 mnemonic of `word_count` words (12 or 24) from
/// operating-system entropy (`getrandom`). The phrase is the ONLY backup of the
/// wallet it seeds — whoever holds it controls the account — so it must be
/// generated privately and never transmitted. Returns the space-separated phrase.
pub fn generate_mnemonic(word_count: usize) -> Result<String, HdError> {
    let entropy_len = match word_count {
        12 => 16,
        24 => 32,
        other => {
            return Err(HdError::Mnemonic(format!(
                "word count must be 12 or 24, got {other}"
            )))
        }
    };
    let mut entropy = [0u8; 32];
    let slice = &mut entropy[..entropy_len];
    getrandom::getrandom(slice).map_err(|e| HdError::Mnemonic(format!("OS entropy: {e}")))?;
    let mnemonic = Mnemonic::from_entropy(slice).map_err(|e| HdError::Mnemonic(e.to_string()))?;
    let phrase = mnemonic.to_string();
    slice.fill(0); // best-effort wipe of the entropy copy
    Ok(phrase)
}

fn hmac_sha512(key: &[u8], data: &[u8]) -> [u8; 64] {
    let mut mac = HmacSha512::new_from_slice(key).expect("HMAC accepts a key of any length");
    mac.update(data);
    let out = mac.finalize().into_bytes();
    let mut bytes = [0u8; 64];
    bytes.copy_from_slice(&out);
    bytes
}

fn split(i: [u8; 64]) -> ([u8; 32], [u8; 32]) {
    let mut key = [0u8; 32];
    let mut chain = [0u8; 32];
    key.copy_from_slice(&i[..32]);
    chain.copy_from_slice(&i[32..]);
    (key, chain)
}

/// SLIP-0010 Ed25519 master: `I = HMAC-SHA512("ed25519 seed", seed)`.
fn master(seed: &[u8]) -> ([u8; 32], [u8; 32]) {
    split(hmac_sha512(ED25519_CURVE, seed))
}

/// SLIP-0010 Ed25519 hardened child: `I = HMAC-SHA512(c_par, 0x00 || k_par ||
/// ser32(index))`; the child key is `I_L` directly (no scalar add).
fn ckd(key: &[u8; 32], chain: &[u8; 32], index: u32) -> ([u8; 32], [u8; 32]) {
    let mut data = [0u8; 1 + 32 + 4];
    data[1..33].copy_from_slice(key);
    data[33..].copy_from_slice(&index.to_be_bytes());
    split(hmac_sha512(chain, &data))
}

/// A hierarchical-deterministic SOV wallet: one BIP-39 seed from which an entire
/// tree of Ed25519 [`Keypair`]s is derived deterministically.
pub struct HdWallet {
    seed: [u8; 64],
}

impl HdWallet {
    /// Build a wallet from a BIP-39 mnemonic, optionally salted by a passphrase
    /// (the "25th word"). Rejects a phrase that fails the wordlist/checksum.
    pub fn from_mnemonic(phrase: &str, passphrase: &str) -> Result<Self, HdError> {
        let mnemonic = Mnemonic::parse(phrase).map_err(|e| HdError::Mnemonic(e.to_string()))?;
        Ok(HdWallet {
            seed: mnemonic.to_seed(passphrase),
        })
    }

    /// Build a wallet directly from a raw 64-byte BIP-39 seed.
    pub fn from_seed(seed: [u8; 64]) -> Self {
        HdWallet { seed }
    }

    /// Derive the raw 32-byte **leaf seed** at a path of indices (each hardened
    /// automatically — SLIP-0010 Ed25519 has no non-hardened derivation). This is
    /// the deterministic, scheme-agnostic HD output: the SDK derives the identical
    /// bytes for the same mnemonic and path. Feed it to whichever key scheme the
    /// account uses — [`Keypair::hybrid_from_seed`] for SOV's default post-quantum
    /// key, or [`Keypair::from_seed`] for a legacy Ed25519 key.
    pub fn derive_seed_at_path(&self, path: &[u32]) -> [u8; 32] {
        let (mut key, mut chain) = master(&self.seed);
        for &index in path {
            let (k, c) = ckd(&key, &chain, index | HARDENED);
            key = k;
            chain = c;
        }
        key
    }

    /// The 32-byte leaf seed for a SOV `account` and address `index` on the
    /// standard BIP-44 path `m/44'/SOV_COIN_TYPE'/account'/0'/index'`.
    pub fn derive_seed(&self, account: u32, index: u32) -> [u8; 32] {
        self.derive_seed_at_path(&[PURPOSE, SOV_COIN_TYPE, account, 0, index])
    }

    /// Derive the **hybrid Ed25519 + ML-DSA-65** keypair (SOV's default,
    /// post-quantum scheme `0x01`) at a path of indices. The same mnemonic always
    /// reproduces the same hybrid key.
    pub fn derive_keypair_at_path(&self, path: &[u32]) -> Keypair {
        Keypair::hybrid_from_seed(self.derive_seed_at_path(path))
    }

    /// Derive the hybrid post-quantum keypair for a SOV `account` and address
    /// `index` on the standard BIP-44 path. Its public key is the value that
    /// authorizes a named account on-chain. This is the scheme new wallets should
    /// use — a 12/24-word phrase restores this exact keypair every time.
    pub fn derive_keypair(&self, account: u32, index: u32) -> Keypair {
        Keypair::hybrid_from_seed(self.derive_seed(account, index))
    }

    /// Derive the legacy classical **Ed25519** keypair (scheme `0x00`) for an
    /// `account`/`index`. The SDK derives the identical key from the same leaf, so
    /// this is the cross-implementation parity anchor; prefer
    /// [`derive_keypair`](Self::derive_keypair) (post-quantum) for new accounts.
    pub fn derive_ed25519_keypair(&self, account: u32, index: u32) -> Keypair {
        Keypair::from_seed(self.derive_seed(account, index))
    }

    /// The raw 64-byte BIP-39 seed (secret material — handle with care).
    pub fn export_seed(&self) -> [u8; 64] {
        self.seed
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    const TREZOR: &str = "abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about";

    #[test]
    fn cross_impl_parity_with_the_sdk() {
        // The SDK (sdk/src/hd.ts) derives this Ed25519 public key for the Trezor
        // reference mnemonic at the standard SOV path m/44'/SOV'/0'/0'/0'. Both
        // implementations derive the identical SLIP-0010 leaf, so the classical
        // key matches byte-for-byte — the cross-impl HD parity anchor.
        let w = HdWallet::from_mnemonic(TREZOR, "").unwrap();
        assert_eq!(
            w.derive_ed25519_keypair(0, 0).public_key().to_hex(),
            "cfedb9535588130e7215859f1346fce339c40c6601196d465fea5c7d43f14464"
        );
    }

    #[test]
    fn hybrid_derivation_is_deterministic_and_post_quantum() {
        // SOV's default: one mnemonic restores the same HYBRID post-quantum key
        // every time — this is what a 12/24-word phrase recovers for a real wallet.
        let a = HdWallet::from_mnemonic(TREZOR, "").unwrap();
        let b = HdWallet::from_mnemonic(TREZOR, "").unwrap();
        let ka = a.derive_keypair(0, 0).public_key();
        assert_eq!(
            ka,
            b.derive_keypair(0, 0).public_key(),
            "a phrase reproduces the same hybrid key every time"
        );
        assert_eq!(
            ka.scheme(),
            "hybrid65",
            "HD default is the post-quantum hybrid"
        );
        // The hybrid key is a distinct scheme from the legacy Ed25519 key derived
        // from the same leaf (independent Blake3 domain-separated components).
        assert_ne!(ka, a.derive_ed25519_keypair(0, 0).public_key());
        // And the underlying leaf is itself deterministic.
        assert_eq!(a.derive_seed(0, 0), b.derive_seed(0, 0));
    }

    #[test]
    fn deterministic_and_distinct_per_account_and_index() {
        let a = HdWallet::from_mnemonic(TREZOR, "").unwrap();
        let b = HdWallet::from_mnemonic(TREZOR, "").unwrap();
        // Deterministic: one mnemonic always yields the same account key.
        assert_eq!(
            a.derive_keypair(0, 0).public_key(),
            b.derive_keypair(0, 0).public_key()
        );
        // Distinct per account and per index.
        let keys = [
            a.derive_keypair(0, 0).public_key(),
            a.derive_keypair(0, 1).public_key(),
            a.derive_keypair(1, 0).public_key(),
        ];
        assert_ne!(keys[0], keys[1]);
        assert_ne!(keys[0], keys[2]);
        assert_ne!(keys[1], keys[2]);
    }

    #[test]
    fn a_passphrase_changes_derivation() {
        let plain = HdWallet::from_mnemonic(TREZOR, "").unwrap();
        let salted = HdWallet::from_mnemonic(TREZOR, "TREZOR").unwrap();
        assert_ne!(plain.export_seed(), salted.export_seed());
    }

    #[test]
    fn a_bad_mnemonic_is_rejected() {
        // Valid words but a broken checksum.
        let bad = "abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon";
        assert!(HdWallet::from_mnemonic(bad, "").is_err());
    }

    #[test]
    fn generated_mnemonics_are_valid_distinct_and_recoverable() {
        let m12 = generate_mnemonic(12).unwrap();
        let m24 = generate_mnemonic(24).unwrap();
        assert_eq!(m12.split_whitespace().count(), 12);
        assert_eq!(m24.split_whitespace().count(), 24);
        // A generated phrase passes BIP-39 validation and seeds a deterministic
        // wallet (so `new` then `import` recover the identical account).
        let a = HdWallet::from_mnemonic(&m24, "").unwrap();
        let b = HdWallet::from_mnemonic(&m24, "").unwrap();
        assert_eq!(
            a.derive_keypair(0, 0).public_key(),
            b.derive_keypair(0, 0).public_key()
        );
        // Real entropy: two generations are overwhelmingly unlikely to collide.
        assert_ne!(
            generate_mnemonic(24).unwrap(),
            generate_mnemonic(24).unwrap()
        );
        // Only 12 or 24 words are accepted.
        assert!(generate_mnemonic(15).is_err());
    }
}
